File: go/problem3.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(LargestPrimeFactor(13195))
}

func Sieve(n int) []int {
	var primes = make([]bool, n+1)
	for i := 1; i <= n; i++ {
		primes[i] = true
	}
	for i := 2; i < n/2; i++ {
		for j := 2; i*j < n; j++ {
			primes[i*j] = false
		}
	}
	var primeNumbers []int
	for i := 1; i <= n; i++ {
		if primes[i] {
			primeNumbers = append(primeNumbers, i)
		}
	}

	return primeNumbers
}

func LargestPrimeFactor(n int) int {
	
	if isPrime(n) {
		return n
	}
  
  var primeList = Sieve(int(math.Sqrt(float64(n))))
	var largestPrimeFactor = 0
	for i := 0; i < len(primeList); i++ {
		if n%primeList[i] == 0 && primeList[i] > largestPrimeFactor {
			largestPrimeFactor = primeList[i]
		}
	}
	return largestPrimeFactor
}

func isPrime(n int) bool {

	if n == 1 {
		return false
	}
	for x := 2; x < int(math.Sqrt(float64(n))); x++ {
		if n%x == 0 {
			return false
		}
	}

	return true
}

File: python/problem3.py
def sieve(n):
    primes = []
    primeNumbers = []

    for x in range(1,n+1):
        primes.append(True)

    for x in range(2,n//2):
        y=2
        while x*y<n:
            primes[x*y]=False
            y +=1

    for x in range(1, n):
        if primes[x]:
            primeNumbers.append(x)

    return primeNumbers

def isPrime(n):
    if n==1: return False
    for x in range(2, int(n ** (1 / 2))+1):
        if n % x == 0: return  False
    return True;

def largestPrimeFactor(n):
    primeList = sieve(int(n ** (1 / 2))+1)
    if isPrime(n): return n

    largestFactor = 1
    for x in primeList:
        if n/x < 0: break
        if n % x == 0 and x > largestFactor:
            largestFactor = x

    return largestFactor

if __name__ == '__main__':
    print(largestPrimeFactor(13195))
